editoast/src/views/infra/queryexpr.rs

```diff
@@ -0,0 +1,564 @@
+use serde_json::Value;
+use std::{
+    collections::{HashMap, VecDeque},
+    rc::Rc,
+};
+
+/// Represents the AST of the query language used in /infra/n/search route.
+///
+/// The following JSON schemes are valid QueryExpr:
+///
+/// * Booleans
+/// * Integers
+/// * Floats
+/// * Strings
+/// * Arrays
+///     * `["column"]`: represents the value of a column (differs from `"column"`
+///        which is just a plain string)
+///     * `["function", arg1, arg2, ...]`: a function call where `argN` are sub-queries
+///
+/// Note that the empty array `[]` and all objects `{...}` are invalid queries.
+///
+/// ```
+/// assert!(QueryExpr::try_from(json!(
+///     ["and", ["=", ["a"], 12],
+///             ["=", ["b"], ["=", true, ["c"]]],
+///             ["like", ["d"], "string"]
+/// ])).is_ok())
+/// ```
+///
+/// See [`eval`] for query evaluation.
+#[derive(Debug, PartialEq)]
+pub enum QueryExpr {
+    Null,
+    Boolean(bool),
+    UnsignedInt(u64),
+    SignedInt(i64),
+    Float(f64),
+    String(String),
+    Column(String),
+    Call(String, Vec<QueryExpr>),
+}
+
+#[derive(Debug, PartialEq)]
+pub struct ExprError(pub String);
+
+impl<S: Into<String>> From<S> for ExprError {
+    fn from(value: S) -> Self {
+        ExprError(value.into())
+    }
+}
```

Comment:
```suggestion
impl<S: AsRef<str>> From<S> for ExprError {
    fn from(value: S) -> Self {
        ExprError(value.as_str().into())
    }
}
```

Reply:
I don't understand how this is better, since we're aiming for a `String` why would you want to use an intermediate `&str` value?  Or is it conventional to do so?